### ISSUE TITLE
Request for localectl to write the X11 keymap to /etc/vconsole.conf

### ISSUE THREAD

**Issue by n3rdopolis** [systemd/systemd#24228]
### Component

other

### Is your feature request related to a problem? Please describe

I hope I am describing this issue correctly, this came up on a draft merge request for Plymouth to make it so it doesn't rely on the kernel mode console for input, but evdev instead.

localectl writes the keymaps to the kernel console in /etc/vconsole.conf, but writes the value for the X11 keymaps in an X11 file, and it seems like it's smart enough to translate the keymap names, As I understand X11 (or user mode) keymap names are not one-to-one with the console keymap names

User mode processes that are not X11, and can't understand X11 config files easily might want a single place to refer to the default configured keymap that libxkbcommon understands.

Having the value saved in  /etc/vconsole.conf by localectl might be good because the file is already known, maybe as an additional key like `XKB_KEYMAP`

### Describe the solution you'd like

It might be hard since now the value will be in two places since one would have to win, (like if someone manually edits the value on vconsole.conf, and not the X11 file, and now the current default one has to be read) but writing the XKB keymap as a key such as `XKB_KEYMAP=foo` to vconsole.conf would allow things that are not Xorg, such as some Wayland servers, or Plymouth. (OK mostly Plymouth) 

Plymouth on distros that use dracut already have /etc/vconsole.conf in their initrd, so amending this file might be the best bet for it

**Comment by marcosfrm** [comment]
Fedora unified kernel console keymaps with XKB long time ago:
https://bugzilla.redhat.com/show_bug.cgi?id=837292
https://src.fedoraproject.org/rpms/kbd/blob/rawhide/f/kbd.spec

**Comment by jwrdegoede** [comment]
> Fedora unified kernel console keymaps with XKB long time ago: https://bugzilla.redhat.com/show_bug.cgi?id=837292 https://src.fedoraproject.org/rpms/kbd/blob/rawhide/f/kbd.spec

Right, Fedora generates console keymaps from xkb keymaps. But what we want to know for the plymouth changes on being worked on is which xkb keymap + variant was used to generate the console keymap. Which AFAICT is currently not available anywhere.

ATM plymouth is using the console keymap since it is using the tty to get keyboard input for unlocking disks encrypted with a password. There are plans: https://gitlab.freedesktop.org/plymouth/plymouth/-/issues/172
to make plymouth use /dev/input/event* directly itself instead of going to the tty.

These plans involve using libxkbcommon to deal with the keymap. The problem is that there is no standard way to get the XKB keymap as set by `localectl set-x11-keymap` the converted/simplified console keymap is stored in /etc/vconsole.conf but going back to the XKB keymap from there (through /usr/share/systemd/kbd-model-map) is somewhat lossy.

Currently the full `LAYOUT [MODEL [VARIANT [OPTIONS]]]` info passed to `localectl set-x11-keymap` is not stored by localectl in either vconsole.conf or locale.conf instead it is only written to `/etc/X11/xorg.conf.d/00-keyboard.conf` in a hard to parse Xorg specific config format; and I guess also in some place for gdm...

So this RFE is about storing the full `LAYOUT [MODEL [VARIANT [OPTIONS]]]` info passed to `localectl set-x11-keymap` in some standard key=value style config file where it is easy for other software to read it.

**Comment by jwrdegoede** [comment]
> makes sense. by all means, send a patch adding this.

Thanks, any suggestions for the format? Since we need to store multiple values when also taking the model / variant and options into account I guess we should probably use multiple key-value pairs for this ?

So instead of adding a KXB_KEYMAP to /etc/vconsole.conf at XKB_LAYOUT XKB_MODEL XKB_VARIANT and XKB_OPTIONS key-value pairs; and then not write the ones not specified by the user to vconsole.conf ?

Also XKB_OPTIONS value needs to be a set of whitespace seperated options I think as I believe there can be more then one option.

**Comment by jwrdegoede** [comment]
>  and then not write the ones not specified by the user to vconsole.conf ?

To be clear what I'm trying to say here is if e.g. the user has only specified a layout then only write XKB_LAYOUT to /etc/vconsole.conf and leave the other ones unset (rather then writing them with an empty value).

**Comment by poettering** [comment]
> Thanks, any suggestions for the format? Since we need to store multiple values when also taking the model / variant and options into account I guess we should probably use multiple key-value pairs for this ?
> 
> So instead of adding a KXB_KEYMAP to /etc/vconsole.conf at XKB_LAYOUT XKB_MODEL XKB_VARIANT and XKB_OPTIONS key-value pairs; and then not write the ones not specified by the user to vconsole.conf ?

yeah

**Comment by poettering** [comment]
> Also XKB_OPTIONS value needs to be a set of whitespace seperated options I think as I believe there can be more then one option.

yeah, if that's how it is it is how it is

**Comment by poettering** [comment]
> To be clear what I'm trying to say here is if e.g. the user has only specified a layout then only write XKB_LAYOUT to /etc/vconsole.conf and leave the other ones unset (rather then writing them with an empty value).

yeah, makes sense

**Comment by Diegovsky** [comment]
> So instead of adding a XKB_KEYMAP to /etc/vconsole.conf at XKB_LAYOUT XKB_MODEL XKB_VARIANT and XKB_OPTIONS key-value pairs; and then not write the ones not specified by the user to vconsole.conf ?
> 
> Also XKB_OPTIONS value needs to be a set of whitespace seperated options I think as I believe there can be more then one option.

I like this. Makes more sense than what I proposed initially

**Comment by n3rdopolis** [comment]
I also like the commas better than spaces, I think that might make it easier for folks that might be `source`ing the file for whatever reason too...

**Comment by legionus** [comment]
> Fedora unified kernel console keymaps with XKB long time ago: https://bugzilla.redhat.com/show_bug.cgi?id=837292 https://src.fedoraproject.org/rpms/kbd/blob/rawhide/f/kbd.spec

FYI I am planning to add xkb support[1] to kbd[2]. If this is done, then xkb keymaps can be used in the linux console with some reservations. I wanted to complete support in the spring, but unfortunately xkbcommon still lacks the necessary APIs. 

[1] https://github.com/legionus/kbd/issues/72
[2] https://github.com/legionus/kbd/tree/xkbcommon

**Comment by Diegovsky** [comment]
I am willing to open a new PR to implement this feature asap

**Comment by poettering** [comment]
See → https://github.com/systemd/systemd/pull/26182, which renames the variables.
